### PR TITLE
Test sorted sets API

### LINKED ISSUE #65: Test sorted sets API
Expand the [ApiSpec](https://github.com/zio/zio-redis/blob/master/src/test/scala/zio/redis/ApiSpec.scala) with sorted sets commands.

### LINKED ISSUE #97: Problem in sorted sets with: zInterStore, zUnionStore, zScan
During creating of unit tests for sorted sets: https://github.com/zio/zio-redis/issues/65
I found out that there are some problems with those 3 commands.
IMO looks like they are not properly "rendered" into redis commands.

I am creating this ticket and will work on it.

### PR DESCRIPTION
**Pull request by maciejbak85**
Resolves #65, resolves #97

Fixed: zInterStore, zUnionStore, zScan
zScore and a problem with INCR in zAdd ( https://github.com/zio/zio-redis/issues/97 )

### CODE REVIEW

**Comment by mijicd** [comment]
@maciejbak85 Don't worry, I've converted it to draft, change its state once you're ready :).

File: redis/src/main/scala/zio/redis/Input.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Input.scala:144]: Let's not introduce this one. I get the idea of calculating `numkeys` automatically, but it's not aligned with the rest of operations we expose.

File: redis/src/main/scala/zio/redis/Output.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Output.scala:166]: What happened here :confused: ?

File: redis/src/main/scala/zio/redis/Output.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Output.scala:172]: Same here.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/api/SortedSets.scala:12]: Please start from `A`. Also, passed like this, any output (even invalid one) will suffice.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/api/SortedSets.scala:90]: This isn't the default order as per the docs.

File: redis/src/main/scala/zio/redis/options/SortedSets.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/options/SortedSets.scala:105]: This should not be needed at all. You can still use `DoubleInput` (not longs) and prepend `WEIGHTS` during command encoding.

**Comment by mijicd** [comment]
@maciejbak85 it looks like you tackled #97 as well, so please mark it in the PR description :pray:

File: redis/src/main/scala/zio/redis/Output.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Output.scala:166]: it just did not work this way ;) Redis returns $1 or $2 here

File: redis/src/main/scala/zio/redis/Output.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Output.scala:172]: Sample redis output looks for example:
```
$2
 10
```

File: redis/src/main/scala/zio/redis/Output.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/Output.scala:166]: The thing is, this is simple string output, it should start with `+`. We have `MultiStringOutput` that already works with `$` afaik.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:90]: hmm will check it

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:12]: indeed, hmm I need to think about some restriction only to String or Long then

File: redis/src/main/scala/zio/redis/Output.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Output.scala:166]: well I did this change based on redis results, what can I say :) otherwise zScore was not working, afaik this StringOutput is not used in other places

File: redis/src/main/scala/zio/redis/Input.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: @mijicd u sure ? as a library user I would expect by lib to make it automatically

File: redis/src/main/scala/zio/redis/Output.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/Output.scala:166]: Oh, you are right. Since this is the case, I propose to use `MultiStringOutput` here, and the do the "rename & nuke" in the follow-up pull request.

File: redis/src/main/scala/zio/redis/Input.scala
Comment (mijicd) [inline redis/src/main/scala/zio/redis/Input.scala:198]: Please no `Seq` :).

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: @mijicd do you have maybe some idea, vision about this ?

File: redis/src/main/scala/zio/redis/Input.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:198]: why not ? ;)

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: ah well now I look at this another time, and dont like this idea ;)
Because each time user will need to be sure about redis output.

zAddWithIncr would be clear enough, about the output type at least.
One method -> one return type

wdyt ?

File: redis/src/main/scala/zio/redis/Input.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/Input.scala:144]: I am sure. It is not aligned with the rest of the API. While it is quite neat, I would rather be consistent, and polish this across the API in the follow-up.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: We shouldn't propagate this to "user land" as it is already determined by presence or absence of increment input. Therefore, let's remove it from here, and determine it during parsing. I think @aleksandarskrbic did something similar in his latest iteration on hashes / geo test.

File: redis/src/main/scala/zio/redis/Input.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/Input.scala:144]: To be perfectly clear, I think the change you propose is way friendlier, and it would be cool if you can open up an issue about it. We can use that issue to summarize all similar problems and get rid of them together.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (mijicd) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: We have a race condition in this conversation :). I'm fine with that approach, but let's see if @aleksandarskrbic and @anovakovic01 can recall the similar situation we had and the solution for it :).

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: ok cool :)

File: redis/src/main/scala/zio/redis/Input.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: ok so for now, user will need to add List size as a param, but we will create new issue for it

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (aleksandarskrbic) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: @maciejbak85 I've stumbled on a similar problem while working on Hash API, specifically HMGET command since we want to return a Map[Field, Value]. Problem is that Redis will return just a list of values. We decided to somehow wire together input and output so while parsing the response, we have information about a command and we can extract fields that users passed as arguments. This was an idea, I haven’t started implementation yet.

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:11]: I usually had situations where I could place everything under a single output type. In this case, you could create sum type which will represent an integer or double output. So something like this:
```scala
sealed trait ZAddReply
object ZAddReply {
  case class Number(value: Long) extends ZAddReply
  case class Score(value: Double) extends ZAddReply
}
```
and then just create `case object ZAddOutput extends Output[ZAddReply]`. In this case, `zAdd` will have `OptionalOutput(ZAddOutput)` as output parameter. WDYT? @maciejbak85 @mijicd

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala]: @anovakovic01 minus I see is that user needs to be aware of what type redis returns, library user should not bother about it

Comment (mijicd) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala]: Well, yes and no, we can parse the output in sort of alternative / orElse fashion. Given possible terms, try them out. Still, it might be cumbersome to propagate.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (anovakovic01) [inline redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: I think that this one should be `OptionalOutput(DoubleOutput)`.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: why ? if you try redis zAdd with INCR, it will return you a string type not double ;)

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: Looks like it returns double as bulk string:
```
If the INCR option is specified, the return value will be Bulk string reply:

The new score of member (a double precision floating point number) represented as string, or nil if the operation was aborted (when called with either the XX or the NX option).
```

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: @anovakovic01 well try in redis console ;)

btw.
https://redis.io/commands/zadd
> If the INCR option is specified, the return value will be Bulk string reply

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: > 127.0.0.1:6379> ZADD set1 1 a 2 b
> (integer) 2
> 127.0.0.1:6379> ZADD set1 INCR 2 a
> "3"

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: I've tried it:
```
ZADD myzset INCR 1 "one" 
$1
1
ZADD myzset INCR 4.2 "two" 
$18
4.2000000000000002
```

Also, on the same link, right below the line that you've just sent it says that it is `a double precision floating point number`, it's just represented as string.

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: hmm what redis version do you have ?

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: ah you are right, I see now implementation of DoubleOutput, wonder why redis-cli is not showimg me those $18 etc. will fix it in a sec

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/api/SortedSets.scala]: > wonder why redis-cli is not showimg me those $18 etc. will fix it in a sec

Try sending requests using telnet, then you should get protocol level responses.

File: redis/src/main/scala/zio/redis/api/SortedSets.scala
Comment (anovakovic01) [inline redis/src/main/scala/zio/redis/api/SortedSets.scala:33]: Just one more change: add `OptionalOutput(DoubleOutput)` to cover cases when output is `$-1\n\r` (which is equivalent to `nil` in redis-cli).
```
ZADD myzset XX INCR 3.2 "three"
$-1

```

File: redis/src/main/scala/zio/redis/Input.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: @mijicd @anovakovic01 
maybe you will know how to fix this:

```
3 more arguments than can be applied to method apply: (a: String, b: Long)(c: String, cs: String*)(d: Option[zio.redis.Aggregate], e: Option[List[Double]])zio.ZIO[zio.redis.RedisExecutor,zio.redis.RedisError,Long] in class Arg2VarargsArg2
            card   <- zInterStore(dest, 2, (first, List(second)), None, Some(List(2, 3)))
```

this happens now when Ive removed NonEmptyListPrecededByCount

```
RedisCommand(
    "ZINTERSTORE",
    Tuple5(
      StringInput,
      LongInput,
      NonEmptyList(StringInput),
      OptionalInput(AggregateInput),
      OptionalInput(WeightsInput)
    ),
    LongOutput
  )
```
some implicts hell ;) trying to investigate it, but its quite hard

**Review by anovakovic01 — APPROVED** [review]
LGTM!

File: redis/src/main/scala/zio/redis/Input.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: I think its a conflict between implicit final class Arg5 and Arg2VarargsArg2

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: @mijicd are you fine with removing Arg2VarargsArg2 ? if its not used and causes problems

Comment (anovakovic01) [reply redis/src/main/scala/zio/redis/Input.scala:144]: @maciejbak85 I think that it's not used anywhere, so feel free to remove it (with this PR every command should be covered by tests).

Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Input.scala:144]: ok removed then

**Comment by maciejbak85** [comment]
@anovakovic01 can you check again PR ? ;) Ive pushed 2 commits after your approve

File: redis/src/main/scala/zio/redis/Output.scala
Comment (anovakovic01) [inline redis/src/main/scala/zio/redis/Output.scala:166]: I think that what @mijicd suggested here is keep the current state from `master` and use `MultiStringOutput` where you need Bulk string.

File: redis/src/main/scala/zio/redis/Output.scala
Comment (anovakovic01) [inline redis/src/main/scala/zio/redis/Output.scala:172]: Same here.

**Review by anovakovic01 — CHANGES_REQUESTED** [review]
@maciejbak85 Sorry, I missed a few things, just fix these, and you have approval. :)

File: redis/src/main/scala/zio/redis/Output.scala
Comment (maciejbak85) [reply redis/src/main/scala/zio/redis/Output.scala:166]: good catch, fixing it

**Review by anovakovic01 — APPROVED** [review]
LGTM!

**Review by mijicd — APPROVED** [review]
Nice work! :rocket:

**Pull request merged**
